### PR TITLE
examples/advanced: add TFLM Fashion-MNIST INT8 demo (tflm_fmnist_demo)

### PR DESCRIPTION
**Pull request by KiyoLelou10**
### Contribution description

This PR adds a new advanced example: **`examples/advanced/tflm_fmnist_demo`**, a minimal **TensorFlow Lite Micro (TFLM)** INT8 inference demo for RIOT.

Highlights:

* **Embedded INT8 model** in `model_data.cpp` (`g_model`, `g_model_len`), distilled from a larger teacher model (see provenance below).
* **48×48 grayscale** Fashion-MNIST sample images are provided. At runtime the app quantizes these images into the model’s input tensor and prints the predicted class.
* Self-contained example showing how to:

  * pull in **`tflite-micro`** via `USEPKG`
  * set up a `MicroInterpreter`, tensor arena, and minimal op resolver (Conv2D, DepthwiseConv2D, AvgPool, FullyConnected, Reshape, Mean, Quantize/Dequantize)
  * embed model bytes and small test inputs in firmware
* Includes helper scripts and ten PNGs to **regenerate** the small test set, but the firmware itself only consumes `samples.h` (the PNGs are for developer convenience).

Provenance of the student model: [https://github.com/KiyoLelou10/FashionMNISTDistill](https://github.com/KiyoLelou10/FashionMNISTDistill)
A larger model was trained and **distilled** into this compact INT8 student. That repo also contains **people images**; with the same pattern shown here, one could train/distill a tiny model to **count people** in a photo (e.g., 0–5+) and run it on RIOT.

Directory (new):

```
examples/advanced/tflm_fmnist_demo/
├─ imgs/                     # 10 PNGs included (already used to build samples.h)
├─ main.cpp                  # RIOT app
├─ Makefile                  # uses C++ and pulls tflite-micro
├─ model_data.cpp            # embedded INT8 TFLite model (C array)
├─ samples.h                 # 10 images (uint8[48*48]) + labels used by firmware
├─ pngs_to_samples_h.py      # optional: rebuild samples.h from imgs/
└─ save_fmnist_10_pngs.py    # optional: re-generate 10 Fashion-MNIST PNGs
```

#### nRF52840 DK

```bash
cd examples/advanced/tflm_fmnist_demo
PORT=/dev/ttyACM0 BOARD=nrf52840dk make all term flash WERROR=0
```

Notes:

* The **first build/flash takes a while** because TFLM packages are compiled (flatbuffers, gemmlowp, ruy, kernels, micro, etc.). This is expected; you’ll see lines like:

  ```
  "make" -C .../tflite-micro/tensorflow/lite/kernels -f .../tflite-kernels.mk
  "make" -C .../tflite-micro/tensorflow/lite/kernels/internal -f .../tflite-kernels-internal.mk
  "make" -C .../tflite-micro/tensorflow/lite/micro -f .../tflite-micro.mk
  ```
* Warnings from TFLM are noisy; `WERROR=0` keeps them non-fatal.
* If `AllocateTensors()` fails on another board, try a larger arena:

  ```
  TENSOR_ARENA_SIZE=131072 PORT=/dev/ttyACM0 BOARD=<your_board> make flash term WERROR=0
  ```

**Expected serial output** (example):

```
[#0] pred=Sneaker     (true=Sandal)      time=3148495 us
[#1] pred=Shirt       (true=Shirt)       time=3148515 us
[#2] pred=Shirt       (true=Coat)        time=3148498 us
[#3] pred=Trouser     (true=Trouser)     time=3148511 us
[#4] pred=Shirt       (true=Coat)        time=3148493 us
[#5] pred=Coat        (true=Coat)        time=3148493 us
[#6] pred=Ankle boot  (true=Ankle boot)  time=3148493 us
[#7] pred=Shirt       (true=Coat)        time=3148492 us
[#8] pred=Sneaker     (true=Bag)         time=3148505 us
[#9] pred=Ankle boot  (true=Ankle boot)  time=3148493 us

Done.
```

(With reference kernels on nRF52840, \~3.1s/inference is normal. Enable CMSIS-NN for faster.)

#### Native (Linux host)

```bash
BOARD=native make -j WERROR=0
make term BOARD=native
```

#### Regenerate the sample header (optional)

The firmware uses `samples.h`. PNGs in `imgs/` are provided so reviewers can rebuild:

```bash
python3 pngs_to_samples_h.py --img-dir ./imgs --out-header samples.h --limit 10
```

(Or create new PNGs with `save_fmnist_10_pngs.py` and regenerate `samples.h`.)

### Issues/PRs references

None. This is a self-contained new example.

### CODE REVIEW

File: examples/advanced/tflm_fmnist_demo/Makefile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+APPLICATION = tflm_fmnist_demo
+BOARD ?= nrf52840dk
+RIOTBASE ?= $(CURDIR)/../../..
+
+FEATURES_REQUIRED += cpp
+USEPKG += tflite-micro
+USEMODULE += xtimer
+
+# ✅ List ALL C++ sources here
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/Makefile]:
```suggestion
# List ALL C++ sources here
```

It's generally good practise not to use emojis in source files, even though with modern editors it's possible. Did ChatGPT generate this?

File: examples/advanced/tflm_fmnist_demo/Makefile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+APPLICATION = tflm_fmnist_demo
+BOARD ?= nrf52840dk
+RIOTBASE ?= $(CURDIR)/../../..
+
+FEATURES_REQUIRED += cpp
+USEPKG += tflite-micro
+USEMODULE += xtimer
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/Makefile]: `xtimer` has been deprecated for a long time, it would be good to use `ztimer` instead.

File: examples/advanced/tflm_fmnist_demo/main.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    // If header size != model size, warn (still works if 48x48 vs model is 48x48)
+    if (in->dims->data[1] != IMG_H || in->dims->data[2] != IMG_W) {
+        printf("WARNING: header image size %dx%d != model expects %dx%d\n",
+               IMG_W, IMG_H, in->dims->data[2], in->dims->data[1]);
+    }
+
+    for (int i = 0; i < SAMPLE_COUNT; ++i) {
+        quantize_into_input(&interpreter, g_images[i]);
+
+        uint32_t t0 = xtimer_now_usec();
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/main.cpp]: Do you really need microsecond precision? On embedded system, that can be quite power consuming.

File: examples/advanced/tflm_fmnist_demo/main.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
+// main.cpp — RIOT + TFLite Micro, runtime quantization from uint8 images (48x48)
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/main.cpp]: Please add a header to your source files according to our Coding Convention: https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#documentation

File: examples/advanced/tflm_fmnist_demo/main.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
+// main.cpp — RIOT + TFLite Micro, runtime quantization from uint8 images (48x48)
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <math.h>
+
+#include "xtimer.h"
+#include "samples.h"    // generated by the Python helper: g_images (uint8), g_labels
+
+// TFLite Micro
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/main.cpp]: RIOT uses C-Style Comments instead of C++ Style comments, please adapt the source files according to our Coding Convention: https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#comments

File: examples/advanced/tflm_fmnist_demo/main.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+// Your quantized model bytes from model_data.cc
+extern "C" {
+extern const unsigned char g_model[];
+extern const unsigned int g_model_len;
+}
+
+// Adjust at build time: make TENSOR_ARENA_SIZE=262144
+#ifndef TENSOR_ARENA_SIZE
+#define TENSOR_ARENA_SIZE (200 * 1024)
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/main.cpp]:
```suggestion
#  define TENSOR_ARENA_SIZE (200 * 1024)
```

Preprocessor directives inside of `#if`s should be indended, please refer to our Coding Convention: https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#indentation-of-preprocessor-directives

File: examples/advanced/tflm_fmnist_demo/Makefile (hunk 1 above)
Comment (KiyoLelou10) [reply examples/advanced/tflm_fmnist_demo/Makefile]: I think so yes, since I had a bug and it gave me a new version of the Makefile, by the way, my work in RIOT was part of a mandatory internship I did for Mr. Hahm at my university which I finished 2 days ago, since I am starting to do my thesis now I do not have a lot of time, I will probably only after completing it, have time to update all of the mentioned aspects.
In the meantime you can gladly go through my code in more depth and point out other flaws if you want. 
Thank you very much for your review thus far, I do not know how high the demand is, but I saw that this tf light micro library was available in RIOT, so I thought it would be good to make an advanced example out of this

File: examples/advanced/tflm_fmnist_demo/main.cpp [hunk 7]
```diff
@@ -0,0 +1,160 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Your Name Andrej Sum-Shik <andrej.sum-shik@stud.fra-uas.de>
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/main.cpp:2]:
```suggestion
 * SPDX-FileCopyrightText: 2025 Andrej Sum-Shik <andrej.sum-shik@stud.fra-uas.de>

```

File: examples/advanced/tflm_fmnist_demo/Makefile.ci [hunk 8]
```diff
@@ -0,0 +1,3 @@
+BOARD_INSUFFICIENT_MEMORY := \
+    msba2 \
+    stk3200
```
Comment (crasbe) [inline examples/advanced/tflm_fmnist_demo/Makefile.ci:3]:
```suggestion
BOARD_INSUFFICIENT_MEMORY := \
    acd52832 \
    adafruit-itsybitsy-m4 \
    adafruit-metro-m4-express \
    adafruit-pybadge \
    airfy-beacon \
    alientek-pandora \
    arduino-due \
    arduino-mkr1000 \
    arduino-mkrfox1200 \
    arduino-mkrwan1300 \
    arduino-mkrzero \
    arduino-nano-33-iot \
    arduino-zero \
    avsextrem \
    b-l072z-lrwan1 \
    b-l475e-iot01a \
    bastwan \
    bitcraze-crazyflie21-main \
    blackpill-stm32f103c8 \
    blackpill-stm32f103cb \
    bluepill-stm32f030c8 \
    bluepill-stm32f103c8 \
    bluepill-stm32f103cb \
    calliope-mini \
    cc1312-launchpad \
    cc1350-launchpad \
    cc1352-launchpad \
    cc1352p-launchpad \
    cc2538dk \
    cc2650-launchpad \
    cc2650stk \
    dwm1001 \
    e104-bt5010a-tb \
    e104-bt5011a-tb \
    e180-zg120b-tb \
    ek-lm4f120xl \
    esp32-ethernet-kit-v1_0 \
    esp32-ethernet-kit-v1_1 \
    esp32-ethernet-kit-v1_2 \
    esp32-heltec-lora32-v2 \
    esp32-mh-et-live-minikit \
    esp32-olimex-evb \
    esp32-ttgo-t-beam \
    esp32-wemos-d1-r32 \
    esp32-wemos-lolin-d32-pro \
    esp32-wroom-32 \
    esp32-wrover-kit \
    esp32c3-devkit \
    esp32c3-wemos-mini \
    esp32h2-devkit \
    esp32s2-devkit \
    esp32s2-lilygo-ttgo-t8 \
    esp32s2-wemos-mini \
    esp8266-esp-12x \
    esp8266-olimex-mod \
    esp8266-sparkfun-thing \
    f4vi1 \
    feather-m0 \
    feather-m0-lora \
    feather-m0-wifi \
    firefly \
    frdm-k22f \
    frdm-kl43z \
    frdm-kw41z \
    generic-cc2538-cc2592-dk \
    hamilton \
    hip-badge \
    i-nucleo-lrwan1 \
    ikea-tradfri \
    im880b \
    iotlab-a8-m3 \
    iotlab-m3 \
    limifrog-v1 \
    lobaro-lorabox \
    lora-e5-dev \
    lsn50 \
    maple-mini \
    mbed_lpc1768 \
    mcb2388 \
    microbit \
    microbit-v2 \
    msba2 \
    msbiot \
    mulle \
    nrf51dk \
    nrf51dongle \
    nrf52832-mdk \
    nrf52dk \
    nucleo-c031c6 \
    nucleo-c071rb \
    nucleo-f030r8 \
    nucleo-f031k6 \
    nucleo-f042k6 \
    nucleo-f070rb \
    nucleo-f072rb \
    nucleo-f091rc \
    nucleo-f103rb \
    nucleo-f207zg \
    nucleo-f302r8 \
    nucleo-f303k8 \
    nucleo-f303re \
    nucleo-f303ze \
    nucleo-f334r8 \
    nucleo-f401re \
    nucleo-f410rb \
    nucleo-f411re \
    nucleo-f429zi \
    nucleo-f439zi \
    nucleo-f446re \
    nucleo-f446ze \
    nucleo-g070rb \
    nucleo-g071rb \
    nucleo-g431rb \
    nucleo-g474re \
    nucleo-l011k4 \
    nucleo-l031k6 \
    nucleo-l053r8 \
    nucleo-l073rz \
    nucleo-l152re \
    nucleo-l412kb \
    nucleo-l432kc \
    nucleo-l433rc \
    nucleo-l452re \
    nucleo-l476rg \
    nucleo-wl55jc \
    nz32-sc151 \
    olimexino-stm32 \
    omote \
    opencm904 \
    openlabs-kw41z-mini \
    openlabs-kw41z-mini-256kib \
    openmote-b \
    openmote-cc2538 \
    p-nucleo-wb55 \
    pba-d-01-kw2x \
    phynode-kw41z \
    pinetime \
    pyboard \
    qn9080dk \
    remote-pa \
    remote-reva \
    remote-revb \
    ruuvitag \
    sam4s-xpro \
    samd10-xmini \
    samd20-xpro \
    samd21-xpro \
    saml10-xpro \
    saml11-xpro \
    saml21-xpro \
    samr21-xpro \
    samr30-xpro \
    samr34-xpro \
    seeedstudio-xiao-esp32c3 \
    seeeduino_arch-pro \
    seeeduino_xiao \
    sensebox_samd21 \
    serpente \
    slstk3400a \
    slstk3401a \
    sltb001a \
    sltb009a \
    slwstk6000b-slwrb4150a \
    slwstk6220a \
    sodaq-autonomo \
    sodaq-explorer \
    sodaq-one \
    sodaq-sara-aff \
    sodaq-sara-sff \
    spark-core \
    stk3200 \
    stk3600 \
    stk3700 \
    stm32c0116-dk \
    stm32c0316-dk \
    stm32f030f4-demo \
    stm32f0discovery \
    stm32f3discovery \
    stm32f429i-disc1 \
    stm32f429i-disco \
    stm32f4discovery \
    stm32f7508-dk \
    stm32g0316-disco \
    stm32l0538-disco \
    stm32l476g-disco \
    stm32mp157c-dk2 \
    teensy31 \
    thingy52 \
    ublox-c030-u201 \
    udoo \
    usb-kw41z \
    weact-f401cc \
    weact-f401ce \
    weact-f411ce \
    weact-g030f6 \
    wemos-zero \
    xg23-pk6068a \
    yarm \
    yunjia-nrf51822 \
    #
```

Comment (crasbe) [reply examples/advanced/tflm_fmnist_demo/Makefile.ci:3]: I ran `make generate-Makefile.ci` on `skyleaf` to get a list of all boards that don't have enough memory... Note that it's most boards we currently support 😅

Comment (maribu) [reply examples/advanced/tflm_fmnist_demo/Makefile.ci:3]: One way to limit the impact on CI would be to limit compilation to just 1-3 boards in the CI and add a warning to the doc that we had to limit build testing here.

There is a good chance that regressions will slip in with limited CI testing, of course. But if the alternative is to not merge an example based on CI times, I'd rather go with limiting CI testing and being upfront about this with a warning in the doc.

Note that I have not looked into the contents of the PR; at this point I'm neither in favor nor opposed to this PR, as I have not looked into it.

**Comment by aabadie** [comment]
I see it a bit redundant with the existing test application (https://github.com/RIOT-OS/RIOT/tree/master/tests/pkg/tflite-micro) which also provide an example to infer MNIST data.

File: examples/advanced/tflm_fmnist_demo/Makefile.ci (hunk 8 above)
Comment (mcr) [reply examples/advanced/tflm_fmnist_demo/Makefile.ci:3]: * I agree with limiting CI for PRs to 1-3 boards.
* the full run needs to be ~weekly.

**Comment by aabadie** [comment]
> I see it a bit redundant with the existing test application (https://github.com/RIOT-OS/RIOT/tree/master/tests/pkg/tflite-micro) which also provide an example to infer MNIST data.

Oh sorry, the model is trained on Fashion-MNIST which is a different dataset. Sorry for the noise

File: examples/advanced/tflm_fmnist_demo/main.cpp (hunk 7 above)
Comment (KiyoLelou10) [reply examples/advanced/tflm_fmnist_demo/main.cpp:2]: Well, I can try to train a much smaller model and then use a smaller model for this example this way, it should be much faster and maybe I can also make it run on more devices. Do you think this would make sense?

**Comment by KiyoLelou10** [comment]
> I see it a bit redundant with the existing test application (https://github.com/RIOT-OS/RIOT/tree/master/tests/pkg/tflite-micro) which also provide an example to infer MNIST data.

Okay thank you for pointing it out, I have actually not seen this example when I made the pr, sorry. Well even though it is FashionMNIST and not MNIST I would understand many other developers making the argument that this is redundant regardlessly. The more striking difference I see to the example is that I have also used distillation, which from what I read is not used in the example. Well as I already see other problems I would adjust this model to be smaller, thus faster and also runnable on more devices.
